### PR TITLE
Exclude modules defined in the current build from dependency checks

### PR DESCRIPTION
**Pull request by julienrf**

### CODE REVIEW

**Comment by julienrf** [comment]
I realized that the new features of release 1.1.0 don’t play well with multi-projects and “intermediate” version numbers (version numbers used in-between releases, such as `1.2.3+4-1234abcd`). Here is a test case that demonstrates the issue.

When checking the compatibility of the dependencies of module `b`, which depends on module `a`, the current version of `a` is `1.0.0+1-...`, which is considered to be incompatible with its previous version, `1.0.0`. Although in practice, that version number is temporary, and eventually the next version number will be `1.0.1` or `1.1.0`.

**Comment by julienrf** [comment]
I guess it would be safe to exclude sub-projects from the dependencies of a project `x`, when checking the compatibility of the dependencies of `x`. But I am not sure how to find such sub-projects. Maybe an sbt guru such as @eed3si9n could help? :pray:

**Comment by eed3si9n** [comment]
Wouldn't dynver-derived automatic version number cause issue with Mima checking in general regardless of the inter-project (internal) dependency? I'd think the way out is to make a fake tag like `1.1.0-alpha1` to indicate the intent of eventual version upgrade.

**Comment by julienrf** [comment]
> Wouldn't dynver-derived automatic version number cause issue with Mima checking in general regardless of the inter-project (internal) dependency?

No, MiMa checking works well AFAICT.

> I'd think the way out is to make a fake tag like `1.1.0-alpha1` to indicate the intent of eventual version upgrade.

Could you elaborate? I’m not sure to follow how it would work.

**Comment by julienrf** [comment]
I have added a commit that excludes the modules defined by the current build definition from the dependency checks. That makes the test pass. Please let me know what you think about it.

**Comment by eed3si9n** [comment]
> Could you elaborate? I’m not sure to follow how it would work.

Let's say you're on version 1.0.1, and you want to add a new method.
As a workaround, you can create a tag `1.1.0-alpha1` on the commit that you do not intent do release, and then add the method. This should force sbt-dynver to generate 1.1.0-alpha-based version name?

> I have added a commit that excludes the modules defined by the current build definition from the dependency checks. That makes the test pass. Please let me know what you think about it.

I guess since internal dependencies typically share the version number that's fine.

In a grand scheme of things, maybe this should be filed as an issue on sbt-dynver so it would generate semver friendly synthetic version number immediately after 1.0.0.

**Comment by julienrf** [comment]
> Let's say you're on version 1.0.1, and you want to add a new method.
As a workaround, you can create a tag 1.1.0-alpha1 on the commit that you do not intent do release, and then add the method. This should force sbt-dynver to generate 1.1.0-alpha-based version name?

That looks to me like a complicated workaround (with the negative side-effect of polluting the Git tags).

**Comment by julienrf** [comment]
> I guess since internal dependencies typically share the version number that's fine.

Yes, but this is not always the case (this might be a niche situation, though…).

There can be a problematic situation like the following, in the same build:

~~~ scala
val a = project.settings(
  versionPolicyIntention := Compatibility.BinaryCompatible,
  version := 1.1.0
)
val b = project.settings(
  versionPolicyIntention := Compatibility.BinaryAndSourceCompatible,
  version := 1.0.1
).dependsOn(a)
~~~

And, checking `b/versionPolicyCheck` would not catch the fact that module `a` will break the compatibility guarantees of `b` (assuming the previous release was `1.0.0` for both projects).

So, I’m tempted to introduce a new setting `versionPolicyIgnoreInternalDependencyVersions` that would let users define a regexp for version numbers of internal dependencies that should be ignored.

~~~
versionPolicyIgnoreInternalDependencyVersions := Some("^[0-9]+\\.[0-9]+\\.[0-9]+\\+[0-9]+") // ignores version numbers like `1.2.3+4`
~~~

File: sbt-version-policy/src/main/scala/sbtversionpolicy/SbtVersionPolicyPlugin.scala [hunk 1]
```diff
@@ -17,6 +19,9 @@ object SbtVersionPolicyPlugin extends AutoPlugin {
     )
 
   object autoImport extends SbtVersionPolicyKeys {
+    // Note: the key is defined here rather than in the trait `SbtVersionPolicyKey` because adding a
+    // member to a trait would break binary compatibility.
+    final val versionPolicyIgnoredInternalDependencyVersions = settingKey[Option[Regex]]("Exclude dependencies to projects of the current build whose version matches this regular expression.")
```
Comment (julienrf) [inline sbt-version-policy/src/main/scala/sbtversionpolicy/SbtVersionPolicyPlugin.scala:24]: I don’t understand why I had to do this hack although sbt-version-policy uses Scala 2.12, which should support adding new methods to traits?

Comment (julienrf) [reply sbt-version-policy/src/main/scala/sbtversionpolicy/SbtVersionPolicyPlugin.scala]: It seems it’s not possible to add a `val` to a trait, but a `def` is OK. See https://github.com/lightbend/mima/blob/main/functional-tests/src/test/trait-added-val-in-new-version-nok/problems.txt

**Comment by nafg** [comment]
IMO ideally sbt-dynver or something like it would take versionPolicyIntention into account.

**Pull request merged**